### DISCUSSION TITLE
Support an equivalent to Mockito session

### DISCUSSION THREAD

**Discussion by antoine-mulet** [mockk/mockk#930] [Ideas]
👋 Hi, would it be possible and desirable to support an equivalent to [MockitoSession](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/MockitoSession.html)?

Of course there's already the verification confirmation in MockK but having to write all the `verify` manually often adds boilerplate to tests in my opinion. Re-using the code displayed in [verification confirmation](https://mockk.io/#verification-confirmation), I would love to simply write

```
val car = mockk<Car>()

every { car.drive(Direction.NORTH) } returns Outcome.OK
every { car.drive(Direction.SOUTH) } returns Outcome.OK

car.drive(Direction.NORTH) // returns OK
car.drive(Direction.SOUTH) // returns OK
```

and the following part would be sort of implicit

```
verify {
    car.drive(Direction.SOUTH)
    car.drive(Direction.NORTH)
}

confirmVerified(car) // makes sure all calls were covered with verification
```

I think it would be a great addition to MockK to help driving cleaner tests by eliminating boilerplate code. What do you think?

**Comment by Raibaz** [comment]
Perhaps [this](https://mockk.io/#automatic-unnecessary-stubbing-check) is similar to what you are referring to?

**Reply by antoine-mulet** [reply]
@Raibaz Thanks for replying so quickly. Actually you're right. Given that MockK already gives a nice exception when a mock is called with incorrect arguments (e.g.`no answer found for: Foo(#1).doSomething(42)`) without having to verify it explicitly, the only thing that is missing compared to when using MockitoSession is checking for unnecessary stubs.

Automatic unnecessary stubbing check is a great feature! I am looking forward to migrating test code from Mockito to MockK now I understand MockK a bit better. Thanks!

**Reply by Raibaz** [reply]
Unnecessary stubbing check was actually added pretty recently exactly for that need :)

Glad it helps.
